### PR TITLE
Switch macOS embedding to proc table embedder API

### PR DESCRIPTION
**Pull request by stuartmorgan**
## Description

Converts the macOS embedding to use the new proc table version of the embedding API, and adds one example unit test using it to demonstrate and validate the approach.

## Related Issues

https://github.com/flutter/flutter/issues/66935

## Tests

I added the following tests: One example macOS engine test demonstrating mocking an embedder method.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Let me know if you'd like me to split the macOS changes out into a separate PR; it's easily separable, but I thought it would be useful to have and end-to-end example for initial review.

**Comment by chinmaygarde** [comment]
> Let me know if you'd like me to split the macOS changes out into a separate PR.

I'd prefer it if you don't mind. I'd be more comfortable evaluating the usability of the PR on its own if the changes for a specific embedder were separated out. You can even just rework this PR into two separate commits if thats easier.

**Comment by stuartmorgan** [comment]
Split out into https://github.com/flutter/engine/pull/21813; I'll make this a draft for now and then rebase it once that one lands.

**Comment by stuartmorgan** [comment]
There's some minor unit test cleanup here; while adding the example test I found that the test file naming was inconsistent within this embedding, so I standardized it to match naming in the iOS embedding.

**Review by chinmaygarde — APPROVED** [review]

File: shell/platform/darwin/macos/BUILD.gn
Comment (cbracken) [inline shell/platform/darwin/macos/BUILD.gn:87]: Looks like this breaks FlutterEngine.mm due to a few symbols defined in the `#infndef FLUTTER_ENGINE_NO_PROTOTYPES` block:
* `FlutterEngineCollectAOTData`
* `FlutterEngineRunsAOTCompiledDartCode`
* `FlutterEngineCreateAOTData`

**Review by cbracken — APPROVED** [review]
lgtm for the change modulo breakage w.r.t. #21813

File: shell/platform/darwin/macos/BUILD.gn
Comment (stuartmorgan) [reply shell/platform/darwin/macos/BUILD.gn:87]: I apparently didn't rebuild locally after rebasing on master 😳 The code that's not compiling is new, so hadn't been updated to use the proc table.

**Pull request merged**
